### ISSUE TITLE
Elasticsearch 5.0.0-rc1 transport-5.0.0-rc1.jar not in download on https://www.elastic.co/downloads/elasticsearch

### ISSUE THREAD

**Issue by toliwa** [elastic/elasticsearch#20926]
**Elasticsearch version**: 5.0.0-rc1

**Plugins installed**: [kibana-5.0.0-rc1, X-Pack]

**JVM version**: 1.8.0_92

**OS version**: Linux

**Description of the problem including expected versus actual behavior**:
The downloadable version of Elasticsearch 5.0.0-rc1 from https://www.elastic.co/downloads/elasticsearch (such as elasticsearch-5.0.0-rc1.zip) does not include transport-5.0.0-rc1.jar in its libs/ folder. 

I was trying to test Elasticsearch 5.0.0-rc1 from java, and included all JARs from elasticsearch-5.0.0-rc1/lib/ in the CLASSPATH. 

The "PreBuiltTransportClient" class was not available. I had to additionally download transport-5.0.0-rc1.jar from http://search.maven.org/#artifactdetails%7Corg.elasticsearch.client%7Ctransport%7C5.0.0-rc1%7Cjar

The documentation on: https://www.elastic.co/guide/en/elasticsearch/client/java-api/5.x/transport-client.html refers to PreBuiltTransportClient to be used in the Java API, so it seems to me that this class should be provided together with the other JARs in the official download, and should not be additionally searched and downloaded from another page. This is useful for cases where plain ant is used as a build system with JARs from the local file system. 

**Steps to reproduce**:
1. Elasticsearch 5.0.0-rc1 from https://www.elastic.co/downloads
2. Include JARs from elasticsearch-5.0.0-rc1/lib/ in the CLASSPATH
3. Try to use the PreBuiltTransportClient as documented on: https://www.elastic.co/guide/en/elasticsearch/client/java-api/5.x/transport-client.html
4. The class PreBuiltTransportClient cannot be found

**Provide logs (if relevant)**:

**Describe the feature**:

**Comment by rjernst** [comment]
> Include JARs from elasticsearch-5.0.0-rc1/lib/ in the CLASSPATH

This is not the correct way to use the transport client. You should depend on the new transport jar (that you linked to) and all of its transitive dependencies (even in 2.x the correct way would have been to depend on the elasticsearch jar in maven, transitively). From ant, you can use something like ivy to handle pulling the jar and its dependencies.

> so it seems to me that this class should be provided together with the other JARs in the official download

The download you refer to is the elasticsearch *server*, not a client to communicate with elasticsearch. For that there is the transport client you already pointed out, and the new [rest client](http://search.maven.org/#artifactdetails%7Corg.elasticsearch.client%7Crest%7C5.0.0-rc1%7Cjar).

**Comment by toliwa** [comment]
Thanks for the clarification. 

Using ivy there seems to be a problem, it downloads 27 JARs and then outputs:

    :: problems summary ::
    :::: ERRORS
            unknown resolver null

            unknown resolver null
            [..]

You can reproduce this behavior with:

    java -jar ivy-2.4.0.jar -retrieve ./[artifact](-[classifier])-[revision].[ext] -dependency org.elasticsearch.client transport 5.0.0-rc1

How can one correctly download the client and all its JARs?

**Comment by rjernst** [comment]
You need quotes around the argument to `-retrieve`, otherwise the shell will muck it up.

**Comment by toliwa** [comment]
Thanks. Additionally, I also had to remove the ``.ivy2`` and ``.ivy2ee`` folders in my ``home`` and re-run the ivy command to make the ERROR messages disappear.

**Comment by toliwa** [comment]
After using the downloaded JARs and following the docs to create a PreBuiltTransportClient object, I noticed that still not all dependencies have been resolved, the following Exception occured:

	2016-10-14 13:56:09,328 INFO org.elasticsearch.plugins.PluginsService: loaded plugin [org.elasticsearch.transport.Netty4Plugin]
	Exception in thread "main" java.lang.NoClassDefFoundError: io/netty/channel/RecvByteBufAllocator
			at org.elasticsearch.transport.Netty4Plugin.getSettings(Netty4Plugin.java:39)
			at org.elasticsearch.plugins.PluginsService.lambda$getPluginSettings$0(PluginsService.java:85)
			at java.util.stream.ReferencePipeline$7$1.accept(ReferencePipeline.java:267)
			at java.util.ArrayList$ArrayListSpliterator.forEachRemaining(ArrayList.java:1374)
			at java.util.stream.AbstractPipeline.copyInto(AbstractPipeline.java:481)
			at java.util.stream.AbstractPipeline.wrapAndCopyInto(AbstractPipeline.java:471)
			at java.util.stream.ReduceOps$ReduceOp.evaluateSequential(ReduceOps.java:708)
			at java.util.stream.AbstractPipeline.evaluate(AbstractPipeline.java:234)
			at java.util.stream.ReferencePipeline.collect(ReferencePipeline.java:499)
			at org.elasticsearch.plugins.PluginsService.getPluginSettings(PluginsService.java:85)
			at org.elasticsearch.client.transport.TransportClient.buildTemplate(TransportClient.java:115)
			at org.elasticsearch.client.transport.TransportClient.<init>(TransportClient.java:228)
			at org.elasticsearch.transport.client.PreBuiltTransportClient.<init>(PreBuiltTransportClient.java:69)
			at org.elasticsearch.transport.client.PreBuiltTransportClient.<init>(PreBuiltTransportClient.java:65)

After resolving this, I re-run the code and this time, a lucene class was missing, see below:

	2016-10-14 14:06:44,553 INFO org.elasticsearch.plugins.PluginsService: loaded plugin [org.elasticsearch.transport.Netty4Plugin]
	Exception in thread "main" java.lang.NoClassDefFoundError: org/apache/lucene/queries/TermsQuery
			at org.elasticsearch.percolator.PercolatorFieldMapper.<clinit>(PercolatorFieldMapper.java:79)
			at org.elasticsearch.percolator.PercolatorPlugin.getSettings(PercolatorPlugin.java:72)
			at org.elasticsearch.plugins.PluginsService.lambda$getPluginSettings$0(PluginsService.java:85)
			at java.util.stream.ReferencePipeline$7$1.accept(ReferencePipeline.java:267)
			at java.util.ArrayList$ArrayListSpliterator.forEachRemaining(ArrayList.java:1374)
			at java.util.stream.AbstractPipeline.copyInto(AbstractPipeline.java:481)
			at java.util.stream.AbstractPipeline.wrapAndCopyInto(AbstractPipeline.java:471)
			at java.util.stream.ReduceOps$ReduceOp.evaluateSequential(ReduceOps.java:708)
			at java.util.stream.AbstractPipeline.evaluate(AbstractPipeline.java:234)
			at java.util.stream.ReferencePipeline.collect(ReferencePipeline.java:499)
			at org.elasticsearch.plugins.PluginsService.getPluginSettings(PluginsService.java:85)
			at org.elasticsearch.client.transport.TransportClient.buildTemplate(TransportClient.java:115)
			at org.elasticsearch.client.transport.TransportClient.<init>(TransportClient.java:228)
			at org.elasticsearch.transport.client.PreBuiltTransportClient.<init>(PreBuiltTransportClient.java:69)
			at org.elasticsearch.transport.client.PreBuiltTransportClient.<init>(PreBuiltTransportClient.java:65)

It looks to me that the client is missing several dependencies, or at least that ivy is not resolving them correctly when downloading it.

**Comment by rjernst** [comment]
Can you list which jars ivy downloaded?

**Comment by toliwa** [comment]
The downloaded JARs are:

	commons-codec-1.10.jar
	commons-logging-1.1.3.jar
	compiler-0.9.3.jar
	elasticsearch-5.0.0-rc1.jar
	HdrHistogram-2.1.6.jar
	hppc-0.7.1.jar
	httpcore-4.4.5.jar
	jackson-core-2.8.1.jar
	jna-4.2.2.jar
	joda-convert-1.2.jar
	joda-time-2.9.4.jar
	jopt-simple-5.0.2.jar
	lang-mustache-client-5.0.0-rc1.jar
	lucene-core-6.2.0.jar
	netty-3.10.6.Final.jar
	netty-common-4.1.5.Final.jar
	percolator-client-5.0.0-rc1.jar
	reindex-client-5.0.0-rc1.jar
	rest-5.0.0-rc1.jar
	securesm-1.1.jar
	snakeyaml-1.15.jar
	t-digest-3.0.jar
	transport-5.0.0-rc1.jar
	transport-javadoc-5.0.0-rc1.jar
	transport-netty3-client-5.0.0-rc1.jar
	transport-netty4-client-5.0.0-rc1.jar
	transport-sources-5.0.0-rc1.jar

The ivy report:

	:: resolution report :: resolve 1034ms :: artifacts dl 15ms
			---------------------------------------------------------------------
			|                  |            modules            ||   artifacts   |
			|       conf       | number| search|dwnlded|evicted|| number|dwnlded|
			---------------------------------------------------------------------
			|      default     |   51  |   0   |   0   |   0   ||   27  |   0   |
			---------------------------------------------------------------------
	:: retrieving :: org.elasticsearch.client#transport-caller
			confs: [default]
			27 artifacts copied, 0 already retrieved (18587kB/45ms)

**Comment by rjernst** [comment]
@toliwa Thank you for opening this issue. I see now that we do indeed still have a bug in how the modules which are published for the transport client do not have the correct dependencies (those which core provides).

**Comment by rjernst** [comment]
Hrm, actually perhaps this is a bug in ivy with handling maven poms? The pom for elasticsearch *does* have `lucene-queries` (which is the jar missing for this particular error). And when ivy prints out what it resolves, it shows lucene queries:

```
:: resolving dependencies :: org.elasticsearch.client#transport-caller;working
	confs: [default]
	found org.elasticsearch.client#transport;5.0.0-rc1 in public
	found org.elasticsearch#elasticsearch;5.0.0-rc1 in public
	found org.apache.lucene#lucene-core;6.2.0 in public
	found org.apache.lucene#lucene-analyzers-common;6.2.0 in public
	found org.apache.lucene#lucene-backward-codecs;6.2.0 in public
	found org.apache.lucene#lucene-grouping;6.2.0 in public
	found org.apache.lucene#lucene-highlighter;6.2.0 in public
	found org.apache.lucene#lucene-join;6.2.0 in public
	found org.apache.lucene#lucene-memory;6.2.0 in public
	found org.apache.lucene#lucene-misc;6.2.0 in public
	found org.apache.lucene#lucene-queries;6.2.0 in public
	found org.apache.lucene#lucene-queryparser;6.2.0 in public
	found org.apache.lucene#lucene-sandbox;6.2.0 in public
	found org.apache.lucene#lucene-spatial;6.2.0 in public
	found org.apache.lucene#lucene-spatial-extras;6.2.0 in public
	found org.apache.lucene#lucene-spatial3d;6.2.0 in public
	found org.apache.lucene#lucene-suggest;6.2.0 in public
	found org.elasticsearch#securesm;1.1 in public
	found net.sf.jopt-simple#jopt-simple;5.0.2 in public
	found com.carrotsearch#hppc;0.7.1 in public
	found joda-time#joda-time;2.9.4 in public
	found org.joda#joda-convert;1.2 in public
	found org.yaml#snakeyaml;1.15 in public
	found com.fasterxml.jackson.core#jackson-core;2.8.1 in public
	found com.fasterxml.jackson.dataformat#jackson-dataformat-smile;2.8.1 in public
	found com.fasterxml.jackson.dataformat#jackson-dataformat-yaml;2.8.1 in public
	found com.fasterxml.jackson.dataformat#jackson-dataformat-cbor;2.8.1 in public
	found com.tdunning#t-digest;3.0 in public
	found org.hdrhistogram#HdrHistogram;2.1.6 in public
	found net.java.dev.jna#jna;4.2.2 in public
	found org.elasticsearch.plugin#transport-netty3-client;5.0.0-rc1 in public
	found io.netty#netty;3.10.6.Final in public
	found org.elasticsearch.plugin#transport-netty4-client;5.0.0-rc1 in public
	found io.netty#netty-buffer;4.1.5.Final in public
	found io.netty#netty-codec;4.1.5.Final in public
	found io.netty#netty-codec-http;4.1.5.Final in public
	found io.netty#netty-common;4.1.5.Final in public
	found io.netty#netty-handler;4.1.5.Final in public
	found io.netty#netty-resolver;4.1.5.Final in public
	found io.netty#netty-transport;4.1.5.Final in public
	found org.elasticsearch.plugin#reindex-client;5.0.0-rc1 in public
	found org.elasticsearch.client#rest;5.0.0-rc1 in public
	found org.apache.httpcomponents#httpclient;4.5.2 in public
	found org.apache.httpcomponents#httpcore;4.4.5 in public
	found org.apache.httpcomponents#httpasyncclient;4.1.2 in public
	found org.apache.httpcomponents#httpcore-nio;4.4.5 in public
	found commons-codec#commons-codec;1.10 in public
	found commons-logging#commons-logging;1.1.3 in public
	found org.elasticsearch.plugin#lang-mustache-client;5.0.0-rc1 in public
	found com.github.spullara.mustache.java#compiler;0.9.3 in public
	found org.elasticsearch.plugin#percolator-client;5.0.0-rc1 in public
```

However, the jars actually downloaded do not include lucene queries:

```
HdrHistogram-2.1.6.jar
commons-codec-1.10.jar
commons-logging-1.1.3.jar
compiler-0.9.3.jar
elasticsearch-5.0.0-rc1.jar
hppc-0.7.1.jar
httpcore-4.4.5.jar
jackson-core-2.8.1.jar
jna-4.2.2.jar
joda-convert-1.2.jar
joda-time-2.9.4.jar
jopt-simple-5.0.2.jar
lang-mustache-client-5.0.0-rc1.jar
lucene-core-6.2.0.jar
netty-3.10.6.Final.jar
netty-common-4.1.5.Final.jar
percolator-client-5.0.0-rc1.jar
reindex-client-5.0.0-rc1.jar
rest-5.0.0-rc1.jar
securesm-1.1.jar
snakeyaml-1.15.jar
t-digest-3.0.jar
transport-5.0.0-rc1.jar
transport-javadoc-5.0.0-rc1.jar
transport-netty3-client-5.0.0-rc1.jar
transport-netty4-client-5.0.0-rc1.jar
transport-sources-5.0.0-rc1.jar
```

I think what happens is ivy sees the exlusions set in other dependencies, and applies them globally. Using maven and gradle do not have this problem, and I'm not really sure what to do here..maybe someone else knows some magic ivy incantation to fix this...

**Comment by dgomezs** [comment]
Looks like it is indeed a problem with ivy. I'm using sbt and the problem has been reported and fixed https://github.com/sbt/sbt/issues/2770 

For those using sbt upgrading to fixes the issue for me 0.13.13-RC3

**Comment by jasontedor** [comment]
Superseded by #21170
